### PR TITLE
ARROW-7684: [Rust] Example Flight client and server for DataFusion

### PR DESCRIPTION
**Pull request by andygrove**
This PR adds DataFusion examples for a Flight client and server where the client can send a SQL query to the server and then receive the results.

I have manually tested with a Java client as well to confirm that it works.

### CODE REVIEW

**Comment by nevi-me** [comment]
Hi @andygrove, please test if my changes work.

**Comment by andygrove** [comment]
Thanks @nevi-me ! Testing with a Java client and the server returns 1 instance of FlightData without error but the client fails with:

```
org.apache.arrow.flight.FlightRuntimeException: CallStatus{code=INTERNAL, cause=null, description='Stream completed without receiving schema.'}
```

**Comment by andygrove** [comment]
I guess my next step here might be to get a Java Flight server returning the same data and use wireshark to compare the two servers.

**Comment by nevi-me** [comment]
> I guess my next step here might be to get a Java Flight server returning the same data and use wireshark to compare the two servers.

I know what the issue is, we're supposed to send the schema first before all the record batches.

**Comment by andygrove** [comment]
@nevi-me It works! I was able to get the batch from a Java client!

**Comment by andygrove** [comment]
I think the next steps for this PR are:

- Fix the release verification issues (new inter-crate dependencies were added)
- Add a flight-client.rs so we can test end to end
- Add a README explaining the examples

I am traveling for the next two days but will pick this up when I can.

**Comment by andygrove** [comment]
@nevi-me The only thing missing now is for the client to read the RecordBatch out of the returned FlightData in the client. Do you think you'll have time to help with that? If not, I can have a go.

**Comment by andygrove** [comment]
Well I managed to get the client parsing the schema and data but it seems pretty hacky at the moment.

**Comment by nevi-me** [comment]
I just saw your messages now, I'm looking at the changes that you've made. We could create something to take the flight stream, and return a `RecordBatchReader`,

File: rust/arrow/src/ipc/reader.rs [hunk 1]
```diff
@@ -663,6 +664,21 @@ impl<R: Read> RecordBatchReader for StreamReader<R> {
     }
 }
 
+pub fn schema_from_bytes(bytes: &[u8]) -> Option<Schema> {
+    let ipc = ipc::get_root_as_message(&bytes[..]);
+    ipc.header_as_schema().map(|schema| fb_to_schema(schema))
+}
+
+pub fn recordbatch_from_bytes(
```
Comment (nevi-me) [inline rust/arrow/src/ipc/reader.rs]: you'd need the `data_header` and `data_body` to read a record batch correctly. See my commit fixing this.

Comment (andygrove) [reply rust/arrow/src/ipc/reader.rs]: well that looks much better! Thanks for doing that.

**Comment by andygrove** [comment]
@nevi-me @paddyhoran any objection to me merging this one?

**Review by nevi-me — APPROVED** [review]
No objection from my side

**Pull request closed without merging**